Create new voice editor

xSabrina opened the pull request:
### Description
Closes #106 

This merges **implement a new version of the feature to create and save voice notes**.
It mainly adjusts the note cardViews (since voice notes contain a seekbar and a play/pause and stop button) and creates a voice note editor fragment containing just a record-by-on-click button.

Here are **some details** on the changes:
- A **permission-request** for recording audio has been added to the BookNotesView before opening the voice-note-editor, if the permission is not given yet. If you deny there is an alertDialog to inform the user why it is necessary (clicking ok is reopening the request, clicking on cancel is just canceling the voice-note-editor-opening).
- The new **VoiceNoteEditorFragment** has just one record-button and a help option in the options-menu (alert dialog at the moment - needs to be adjusted to the way we implement our help in our app in general). Clicking on the record-button will light up the button icon slightly and starts a ripple animation as indicator for recording. The record can be stopped either by clicking on the record-button again or leave the fragment (there is a toast as "note has been saved"-feedback).
- The **note-cardView** for our note-lists contains a voice-note related part now for handling the **seekBar and play/pause and stop button**, which is only shown if the note is one of the voice types. The seekBar of an item is usable as soon as the linked recording has been started at least (means: the item's play-button has to be used first) to avoid some bugs - this is indicated by the grayed out seekBar appearance. The item's stop-Button is also just usable if the linked recording has been played at least (so you can not stop any item's playing by pressing another one's stop-button).
- A custom **SeekBarListener** has been added to perform seekBar-features (means: scrolling through the recording by moving the thumb).
- **Methods to fetch a path-string of a noteFile** from the database in NoteModel and NoteDao have been added.
- Also adjusted the **NoteRecyclerViewAdapter params** to contain a NoteModel as well to perform database-related features (like playing notes etc.) (its **methods** have also been adjusted to perfom voice note-features of course).
- Voice note related **play, pause and stop icons** have been added.
- Voice note related **string ressources and a library-key** have been added.
- Added implementations for the seekbar and the ripple-animation to the **build.gradle** and cleaned it up as well (adjusted two implemenation version declarations).

I also did some **general small note-related clean ups**:
- **Typos** in the TextNoteEditorFragment have been fixed.
- **A bug in editting TextNotes** caused by a missing bookId (since the adapter is not longer handling Note-objects, but NoteItem-Objects) in the NoteItems has been fixed by adding a bookId-param to constructors.
- **Unnecessary and legacy constructors** have been removed from some NoteItem-classes.
- The **note-type related microphone.png has been increased in size** since it was the same size as the document.png, but had more margin.
- **fragment_book_notes.xml** has been cleaned up (e.g. replacing dp by sp-units for text-size declarations).

### Affects
Mainly note-related features like the BookNotesView and the database as well, so @SarahKurek should take a look. Also this is just an basic design idea for the voice note features - there might be some things that need or could be adjusted, which @LuisMossburger should take note of.

### Notes for Reviewer
Some suggestions for your testing:
- If you use the emulator: open your extended controls and activate the option **"Virtual microphone uses host audio input"** so you can use your PC microphone.
- **Android only allows to ask a few times for a request by system**. This means if the user denies the request a few times, android will not allow to ask for it again and the request will not be shown anymore and the voice notes can not be used.

Please review & comment/approve. Thanks!

LuisMossburger commented on BibBuddy/app/src/main/java/de/bibbuddy/BookNotesView.java:
```diff
@@ -35,17 +40,43 @@
+    setupPermissionLauncher();
+  }
+
+  /* Register permissions callback, which handles the user's response to the
+   system permissions dialog. Save the return value, an instance of
+   ActivityResultLauncher, as an instance variable.
+   */
```
Proper formatting for the comment :)
```suggestion
  /** Register permissions callback, which handles the user's response to the
    * system permissions dialog. Save the return value, an instance of
    * ActivityResultLauncher, as an instance variable.
    */
```

LuisMossburger commented on BibBuddy/app/src/main/java/de/bibbuddy/BookNotesView.java:
```diff
@@ -35,17 +40,43 @@
+   system permissions dialog. Save the return value, an instance of
+   ActivityResultLauncher, as an instance variable.
+   */
+  private void setupPermissionLauncher() {
+    requestPermissionLauncher =
+        registerForActivityResult(new ActivityResultContracts.RequestPermission(), isGranted -> {
+          if (isGranted) {
+            Bundle bundle = new Bundle();
+            bundle.putLong(LibraryKeys.BOOK_ID, bookId);
+            VoiceNoteEditorFragment nextFrag = new VoiceNoteEditorFragment();
+            nextFrag.setArguments(bundle);
+            requireActivity().getSupportFragmentManager().beginTransaction()
+                .replace(R.id.fragment_container_view, nextFrag,
+                    LibraryKeys.FRAGMENT_VOICE_NOTE_EDITOR)
+                .addToBackStack(null)
+                .commit();
+          }
+        });
+  }
```
A few line breaks for better readability :)

```suggestion
  private void setupPermissionLauncher() {
    requestPermissionLauncher =
        registerForActivityResult(new ActivityResultContracts.RequestPermission(), isGranted -> {
        
          if (isGranted) {
            Bundle bundle = new Bundle();
            bundle.putLong(LibraryKeys.BOOK_ID, bookId);
            VoiceNoteEditorFragment nextFrag = new VoiceNoteEditorFragment();
            nextFrag.setArguments(bundle);
            
            requireActivity().getSupportFragmentManager().beginTransaction()
                .replace(R.id.fragment_container_view, nextFrag,
                    LibraryKeys.FRAGMENT_VOICE_NOTE_EDITOR)
                .addToBackStack(null)
                .commit();
          }
        });
  }
```

LuisMossburger commented on BibBuddy/app/src/main/java/de/bibbuddy/BookNotesView.java:
```diff
@@ -279,42 +283,64 @@ private void setupAddButton() {
-
-
-        return true;
+    pm.setOnMenuItemClickListener(item -> {
+      if (item.getItemId() == R.id.add_text_note) {
+        Bundle bundle = new Bundle();
+        bundle.putLong(LibraryKeys.BOOK_ID, bookId);
+        TextNoteEditorFragment textFrag = new TextNoteEditorFragment();
+        textFrag.setArguments(bundle);
+        requireActivity().getSupportFragmentManager().beginTransaction()
+            .replace(R.id.fragment_container_view, textFrag,
+                LibraryKeys.FRAGMENT_TEXT_NOTE_EDITOR)
+            .addToBackStack(null)
+            .commit();
```
A few new lines for better readability :)

```suggestion
    pm.setOnMenuItemClickListener(item -> {
    
      if (item.getItemId() == R.id.add_text_note) {
        Bundle bundle = new Bundle();
        bundle.putLong(LibraryKeys.BOOK_ID, bookId);
        TextNoteEditorFragment textFrag = new TextNoteEditorFragment();
        textFrag.setArguments(bundle);
        
        requireActivity().getSupportFragmentManager().beginTransaction()
            .replace(R.id.fragment_container_view, textFrag,
                LibraryKeys.FRAGMENT_TEXT_NOTE_EDITOR)
            .addToBackStack(null)
            .commit();
            
```

LuisMossburger commented on BibBuddy/app/src/main/java/de/bibbuddy/BookNotesView.java:
```diff
@@ -279,42 +283,64 @@ private void setupAddButton() {
-        pm.show();
-      }
-    });
+  private void checkRecordPermission() {
+    if (ContextCompat.checkSelfPermission(
+        context, Manifest.permission.RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED) {
+      Bundle bundle = new Bundle();
+      bundle.putLong(LibraryKeys.BOOK_ID, bookId);
+      VoiceNoteEditorFragment voiceFrag = new VoiceNoteEditorFragment();
+      voiceFrag.setArguments(bundle);
+      requireActivity().getSupportFragmentManager().beginTransaction()
+          .replace(R.id.fragment_container_view, voiceFrag,
+              LibraryKeys.FRAGMENT_VOICE_NOTE_EDITOR)
+          .addToBackStack(null)
+          .commit();
+    } else if (shouldShowRequestPermissionRationale(Manifest.permission.RECORD_AUDIO)) {
+      showAudioRecordRequest();
+    } else {
+      requestPermissionLauncher.launch(
+          Manifest.permission.RECORD_AUDIO);
+    }
+  }
```
A few newlines for better readability :)

```suggestion
  private void checkRecordPermission() {
    if (ContextCompat.checkSelfPermission(
        context, Manifest.permission.RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED) {
        
      Bundle bundle = new Bundle();
      bundle.putLong(LibraryKeys.BOOK_ID, bookId);
      VoiceNoteEditorFragment voiceFrag = new VoiceNoteEditorFragment();
      voiceFrag.setArguments(bundle);
      
      requireActivity().getSupportFragmentManager().beginTransaction()
          .replace(R.id.fragment_container_view, voiceFrag,
              LibraryKeys.FRAGMENT_VOICE_NOTE_EDITOR)
          .addToBackStack(null)
          .commit();
          
    } else if (shouldShowRequestPermissionRationale(Manifest.permission.RECORD_AUDIO)) {
      showAudioRecordRequest();
      
    } else {
      requestPermissionLauncher.launch(
          Manifest.permission.RECORD_AUDIO);
    }
  }
```

LuisMossburger commented on BibBuddy/app/src/main/java/de/bibbuddy/NoteModel.java:
```diff
@@ -75,6 +75,25 @@ public Note getLastNote() {
+   *
+   * @return returns the list of voice type note objects.
+   */
+  public List<Note> getVoiceNoteList() {
+    List<Note> noteList = noteDao.findAll();
+    for (int i = 0; i < noteList.size(); i++) {
+      if (noteList.get(i).getType() != 1) {
+        noteList.remove(noteList.get(i));
+      }
+    }
+    return noteList;
+  }
```
Would add newlines here, may be subjective.

```suggestion
  public List<Note> getVoiceNoteList() {
    List<Note> noteList = noteDao.findAll();
    
    for (int i = 0; i < noteList.size(); i++) {
      if (noteList.get(i).getType() != 1) {
        noteList.remove(noteList.get(i));
      }
    }
    
    return noteList;
  }
```

LuisMossburger commented on BibBuddy/app/src/main/java/de/bibbuddy/NoteRecyclerViewAdapter.java:
```diff
@@ -24,18 +34,34 @@
     extends RecyclerView.Adapter<NoteRecyclerViewAdapter.NotesViewHolder> {
 
   private final MainActivity activity;
+  private final NoteModel noteModel;
+  private final ArrayList<MediaPlayer> mediaPlayers;
+  private final ArrayList<ImageButton> playButtons;
+  private final ArrayList<ImageButton> stopButtons;
+  private final ArrayList<ProgressBar> progressBars;
+  private final ArrayList<SeekBarListener> seekBarListeners;
   private List<NoteItem> noteList;
   private ViewGroup parent;
+  private int mediaPlayerPosition;
+  private boolean paused;
```
Maybe some grouping?

```suggestion
  private final MainActivity activity;
  private final NoteModel noteModel;
  
  private final ArrayList<MediaPlayer> mediaPlayers;
  private final ArrayList<ImageButton> playButtons;
  private final ArrayList<ImageButton> stopButtons;
  private final ArrayList<ProgressBar> progressBars;
  private final ArrayList<SeekBarListener> seekBarListeners;
  
  private List<NoteItem> noteList;
  private ViewGroup parent;
  private int mediaPlayerPosition;
  private boolean paused;
```

LuisMossburger commented on BibBuddy/app/src/main/java/de/bibbuddy/NoteRecyclerViewAdapter.java:
```diff
@@ -95,17 +127,139 @@ public void onBindViewHolder(@NonNull NotesViewHolder holder, int position) {
+      paused = false;
+    });
+
+    playButton.setOnClickListener(v -> {
+      ImageButton button = (ImageButton) v;
+      if (v.isSelected()) {
+        if (paused) {
+          mediaPlayerPosition = mediaPlayer.getCurrentPosition();
+          mediaPlayer.seekTo(mediaPlayerPosition);
+          mediaPlayer.start();
+          button.setImageResource(R.drawable.icon_pause);
+        } else {
+          mediaPlayer.pause();
+          button.setImageResource(R.drawable.icon_play);
+        }
+        paused = !paused;
+      } else {
+        resetPlayers();
+        startAudio(mediaPlayer, button, noteItem, seekBarListener);
+        stopButton.setClickable(true);
+      }
+      progressBar.setEnabled(true);
+    });
```
Some new lines here?

```suggestion
    playButton.setOnClickListener(v -> {
      ImageButton button = (ImageButton) v;
      
      if (v.isSelected()) {
        if (paused) {
          mediaPlayerPosition = mediaPlayer.getCurrentPosition();
          mediaPlayer.seekTo(mediaPlayerPosition);
          mediaPlayer.start();
          button.setImageResource(R.drawable.icon_pause);
          
        } else {
          mediaPlayer.pause();
          button.setImageResource(R.drawable.icon_play);
        }
        
        paused = !paused;
      } else {
        resetPlayers();
        startAudio(mediaPlayer, button, noteItem, seekBarListener);
        stopButton.setClickable(true);
      }
      
      progressBar.setEnabled(true);
    });
```

LuisMossburger commented on BibBuddy/app/src/main/java/de/bibbuddy/SeekBarListener.java:
```diff
@@ -0,0 +1,102 @@
+  @Override
+  public void onStartTrackingTouch(SeekBar seekBar) {
+    //Auto-generated method stub
+
```
```suggestion
```

LuisMossburger commented on BibBuddy/app/src/main/res/layout/fragment_voice_note_editor.xml:
```diff
@@ -0,0 +1,42 @@
+
+</RelativeLayout>
+
+
+
+
+
```
```suggestion
```

LuisMossburger commented on BibBuddy/app/src/main/res/layout/list_view_item_note.xml:
```diff
@@ -16,30 +17,88 @@
-                android:contentDescription="@string/description_text_note"
-                android:src="@drawable/document" />
+                android:layout_marginStart="7dp"
+                android:layout_toEndOf="@+id/seekBar"
+                android:background="@null"
+                android:contentDescription="@string/play_button_note"
+                android:scaleType="fitXY"
```
I would suggest using "fitCenter" here - "fitXY" leads to distortions on my screen.

LuisMossburger commented on BibBuddy/app/src/main/res/layout/list_view_item_note.xml:
```diff
@@ -16,30 +17,88 @@
+                android:layout_toEndOf="@+id/play_note"
+                android:background="@null"
+                android:contentDescription="@string/stop_button_note"
+                android:scaleType="fitXY"
```
I would suggest using "fitCenter" here - "fitXY" leads to distortions on my screen.

xSabrina replied:
Thanks - I applied it to all the fitXY-usages.

LuisMossburger commented on BibBuddy/app/src/main/res/layout/list_view_item_note.xml:
```diff
@@ -16,30 +17,88 @@
 
     </RelativeLayout>
-</androidx.cardview.widget.CardView>
+
+</androidx.cardview.widget.CardView>
```
In general, this layout seems to cause problems. 

On both my emulator and my smartphone, play/pause button are distorted and too small and too far right (on my smartphone the buttons didnt show up at all). I would suggest using a ConstraintLayout and constraining from right to left to adjust the size of the bar so that the controls are always visible and big enough. You can have a look at the header layout, where i sized the header text based on the width of elements on its right.

LuisMossburger replied:
Also pls add a newline at the end of the file, github says :)

xSabrina replied:
Thanks for your bug-report: I replaced the relative voice note layout with a constraint layout and adjusted all child-views as suggested.
I added a new line at the end of the file as well.

LuisMossburger commented on BibBuddy/app/src/main/res/menu/fragment_voice_note_editor_menu.xml:
```diff
@@ -0,0 +1,8 @@
+        android:id="@+id/menu_help_voice_editor"
+        android:title="@string/help" />
+
+</menu>
```
```suggestion
</menu>

```

Newline at the end of the file :)

LuisMossburger commented on BibBuddy/app/src/main/res/values/strings.xml:
```diff
@@ -504,4 +520,19 @@
+    <string name="voice_note_saved">Die Sprachnotiz wurde gespeichert.</string>
+
+    <!-- Voice Note Editor Image Descriptions -->
+    <string name="record_microphone">Record Microphone Icon</string>
+    <string name="stop_button_note">Stop Button</string>
+    <string name="play_button_note">Play Button</string>
+    <string name="microphone">Microphone Icon</string>
```
What are these? If these are sole descriptions for icons, please move them up in the respective section and rename them accordingly. Also, please take some time to add more detail to your description for visually impaired.

Please also translate to German :)

xSabrina replied:
You are right - I moved them to the general Image Descriptions section and rewrote the content to be more specific and translated it to german.

LuisMossburger replied:
Great, thanks!

LuisMossburger reviewed (changes requested):
Hey, thanks for the work!

The voice editor works pretty good and I'm excited to have this in the app! :) Some suggestions for code refactoring, as well as the layout for the list item. Also, please use these icons for the text notes controls and the voice note editor. I probably need to create one for the list view myself.

Cheers!

![stop](https://user-images.githubusercontent.com/44579990/112877770-72921500-90c7-11eb-8146-e34b45eeacb3.png)
![pause](https://user-images.githubusercontent.com/44579990/112877785-7756c900-90c7-11eb-8a15-c499480d7321.png)
![right-arrow](https://user-images.githubusercontent.com/44579990/112877790-79208c80-90c7-11eb-9718-759e08a22578.png)
![microphone](https://user-images.githubusercontent.com/44579990/112877799-7a51b980-90c7-11eb-8201-7e9e2ef7e816.png)
https://www.flaticon.com/packs/ui-interface-25/3?word=ui%20interface

xSabrina commented:
> The voice editor works pretty good and I'm excited to have this in the app! :) 

Thanks, same here :)

> Some suggestions for code refactoring, as well as the layout for the list item.
>Also, please use these icons for the text notes controls and the voice note editor. I probably need to create one for the list view myself.

I applied all of your suggestions, implemented your new icons and deleted the old ones - please have another look :)

LuisMossburger reviewed (changes requested):
Hey! Thanks for implementing the suggestions. However, the layout in the list view still seems to cause problems (see screenshot). If you want, I can have another look and try to fix this tomorrow..

![grafik](https://user-images.githubusercontent.com/44579990/112887508-86437880-90d3-11eb-8acc-0fa497eb058d.png)

xSabrina commented:
> Hey! Thanks for implementing the suggestions. However, the layout in the list view still seems to cause problems (see screenshot). If you want, I can have another look and try to fix this tomorrow..

I see - that would be cool actually since the layout looks fine on my devices it is hard to check if the changes fixed it. The layout on your screenshot looks strange however since the voice elements should be constrainted right and next to each other now.

LuisMossburger commented:
Hey!
I made some adjustments and think the layout works properly now. Basically, I wrapped the whole list item in a ConstraintLayout and added constrainedWidth="true" to make it resizeable based on the surroundings. I also adjusted the voice editor button to make it fit with the rest of the apps design. Please have a look and merge if you agree :)

xSabrina commented:
> Hey!
> I made some adjustments and think the layout works properly now. Basically, I wrapped the whole list item in a ConstraintLayout and added constrainedWidth="true" to make it resizeable based on the surroundings.

Hey, thanks for your effort! I see why this is solving the issue.

> I also adjusted the voice editor button to make it fit with the rest of the apps design.

Ah, thank you - that is much better!

> Also, for me the Toast after creating a voice note is not showing up anymore, is that desired behaviour?

The toast works properly for me:

![voice_toast](https://user-images.githubusercontent.com/41908872/112980275-f694dd00-9159-11eb-933d-aa25288a57a9.png)

But there was a time after our first bigger design rework when all of my toast did not show up anymore. After wiping and reinstalling toasts worked again - maybe this might solve your problem as well.

LuisMossburger reviewed (approved):
Alright, ready to merge now!

Will have a look at Toasts in general in my "Help" PR anyways :)
